### PR TITLE
Add new hook and make temporary contracts directory configurable

### PR DESCRIPTION
**Pull request by rynobey**
This PR adds two small features:

- A new hook that gets called right before compiling the instrumented source code (onPreCompile)
- The ability to specify an alternative temporary directory to use for the instrumented source code

I have made an attempt to update the documentation and added the hook to the existing test - please let me know if I can add anything else to make the PR more complete.

### CODE REVIEW

File: plugins/resources/plugin.utils.js [hunk 1]
```diff
@@ -94,7 +94,7 @@ function toRelativePath(pathToFile, pathToParent){
 function getTempLocations(config){
   const contractsRoot = path.parse(config.contractsDir).dir
   const cwd = config.workingDir;
-  const contractsDirName = '.coverage_contracts';
+  const contractsDirName = config.coverage_contracts_temp || '.coverage_contracts';
```
Comment (cgewecke) [inline plugins/resources/plugin.utils.js:97]:
```suggestion
  const contractsDirName = config.coverageContractsTemp || '.coverage_contracts';
```

Could we make this camel-cased, like the other config options? (Will need to be reflected in the docs as well...)

**Review by cgewecke — COMMENTED** [review]
This is great, thanks so much! 

Would it be possible to add docs for the hook and the new temp contracts directory option to the config options table in the 
+ README.md
+ HARDHAT_README.md

A brief sentence about how the preCompile hook is useful would also be nice.  

(That table is a bit painful to add to in markdown, sorry about this...)

File: plugins/resources/plugin.utils.js (hunk 1 above)
Comment (rynobey) [reply plugins/resources/plugin.utils.js:97]: Sure, will update and add to the docs

**Review by cgewecke — APPROVED** [review]
Fantastic, thanks again.

(Will hopefully publish this weekend)

**Comment by rynobey** [comment]
Hi @cgewecke, just wanted to follow up - I could be missing it but I don't see any new releases containing this work?

**Comment by cgewecke** [comment]
Hi! I'm so sorry. I thought I had published this, thanks for the ping. 

I'need to add some tests for #667 tomorrow and will merge and publish then.

**Comment by rynobey** [comment]
Awesome :) Thanks!!

**Comment by rynobey** [comment]
Awesome!

**Pull request merged**